### PR TITLE
Fs 4643 numbering

### PR DESCRIPTION
**Pull request by srh-sloan**
### Change description
Allows the reordering of sections in a round and forms in a section. Also corrects the ordering of sections and forms when one is removed or cloned

- [x] Unit tests and other appropriate tests added or updated
- [x] README and other documentation has been updated / added (if needed)
- [x] Commit messages are meaningful and follow good commit message guidelines (e.g. "FS-XXXX: Add margin to nav items preventing overlapping of logo")

### Screenshots of UI changes (if applicable)
Added move up/move down links to build application screen
Screenshot 2024-09-25 at 12 26 57

Added move up/down links to section screen
Screenshot 2024-09-25 at 12 27 39

### CODE REVIEW

File: app/db/queries/application.py [hunk 1]
```diff
@@ -545,3 +610,136 @@ def _delete_all_components_in_pages(page_ids):
...
+        section_index_to_move_down (int): Current Section.index value of the section to move
+    """
+    round: Round = get_round_by_id(round_id)
+    if section_index_to_move_down >= len(round.sections):
+        raise IndexError("Cannot move the last section down")
+    list_index_to_move_down = section_index_to_move_down - 1  # Need the 0-based index inside the list
+    # Remove the seciton we are moving from the list
+    moved_section = round.sections.pop(list_index_to_move_down)
+    # Reinsert it at the new position - because the list index is 0-based, this equates to the old Section.index
+    round.sections.insert(section_index_to_move_down, moved_section)
```
Comment (ksp37-dluhc) [inline app/db/queries/application.py:645]: Am I right in thinking this is just an element swap? If so, can use bit of Python syntactic sugar :
```
round.sections[list_index_to_move_down],  round.sections[list_index_to_move_down+1] = round.sections[list_index_to_move_down+1], round.sections[list_index_to_move_down]
```

File: app/blueprints/fund_builder/templates/build_application.html [hunk 2]
```diff
@@ -35,8 +35,14 @@ <h2 class="govuk-heading-m">{{ fund.short_name }} - {{round.title_json["en"]}} (
                     <h3 class="govuk-heading-m">{{ section.index }}. {{ section.name_in_apply_json["en"] }}</h3>
                 </span>
                 <span class="app-task-list__task-actions">
-                    <a href='{{ url_for("build_fund_bp.section", round_id=round.round_id, section_id=section.section_id) }}'>Edit</a>
-                    <a href='{{ url_for("build_fund_bp.section",round_id=round.round_id, section_id=section.section_id, action="remove") }}'>Remove</a>
+                    <a class="govuk-link--no-visited-state" href='{{ url_for("build_fund_bp.section", round_id=round.round_id, section_id=section.section_id) }}'>Edit</a>&nbsp;
```
Comment (nuwan-samarasinghe) [inline app/blueprints/fund_builder/templates/build_application.html:38]: To give space cant we use CSS ? may be

File: app/blueprints/fund_builder/templates/section.html [hunk 3]
```diff
@@ -4,49 +4,64 @@
...
-                    {{ form.hidden_tag()}}
-                    {{input(form.name_in_apply_en)}}
-
-                    {{ govukButton({
-                        "text": "Save"
-                        }) }}
+                    <div class="govuk-form-group">
+                        <h3 class="govuk-heading-m">Section Name</h3>
+                        {{ form.hidden_tag() }}
+                        {{ input(form.name_in_apply_en) }}
```
Comment (nuwan-samarasinghe) [inline app/blueprints/fund_builder/templates/section.html:19]: For the inputs cant we use govukInput like in front end ?

File: app/db/queries/application.py (hunk 1 above)
Comment (srh-sloan) [reply app/db/queries/application.py:645]: Wow... mind blown... 🤯 Thank you, I will have a play with this!

File: app/blueprints/fund_builder/templates/build_application.html (hunk 2 above)
Comment (srh-sloan) [reply app/blueprints/fund_builder/templates/build_application.html:38]: Probably yes in future. But we don't have proper designs for these screens yet, they are just POC, so when we have proper designs we can spend the time on better CSS 👍

**Review by nuwan-samarasinghe — APPROVED** [review]
LGTM

File: app/blueprints/fund_builder/routes.py [hunk 4]
```diff
@@ -72,8 +76,15 @@ def section(round_id):
         "round_id": str(round_id),
     }
     existing_section = None
+    # TODO there must be a better way than a pile of ifs...
     if request.args.get("action") == "remove":
```
Comment (ksp37-dluhc) [inline app/blueprints/fund_builder/routes.py:80]: I have no strong feelings about this, but if you're interested, the pattern I tend to use is 
```
action_handlers = {
"remove" : lambda : delete_section_from_round(round_id=round_id, section_id=request.args.get("section_id"), cascade=True),
"move_up" : lambda : move_section_up(round_id=round_id, section_index_to_move_up=int(request.args.get("index"))),
...
}

try:
    action_handlers[request.args.get("action")]()
    return redirect(url_for("build_fund_bp.build_application", round_id=round_id))
except KeyError:
    # Some default handling or pass
```
The one benefit of the above is that the multiple ifs aren't executed (although the same could be achieved with if/elif etc). I'm not sure one is tidier than the other though!

**Review by ksp37-dluhc — APPROVED** [review]
LGTM!

**Pull request merged**